pkg/sql/backfill/mvcc_index_merger.go

```diff
@@ -0,0 +1,325 @@
+				case <-stopProgress:
+					return nil
+				case <-tick.C:
+					pushProgress()
+				}
+			}
+		})
+
+		g.GoCtx(func(ctx context.Context) error {
+			defer close(stopProgress)
+
+			// TODO(rui): some room for improvement on single threaded
+			// implementation, e.g. run merge for spec spans in parallel.
+			for i := range ibm.spec.Spans {
+				sp := ibm.spec.Spans[i]
+				idx := ibm.spec.SpanIdx[i]
+
+				key := sp.Key
+				for key != nil {
+					nextKey, err := ibm.Merge(ctx, ibm.evalCtx.Codec, ibm.desc, ibm.spec.TemporaryIndexes[idx], ibm.spec.AddedIndexes[idx],
+						key, sp.EndKey, ibm.spec.ChunkSize)
+					if err != nil {
+						return err
+					}
+
+					completedSpan := roachpb.Span{}
+					if nextKey == nil {
+						completedSpan.Key = key
+						completedSpan.EndKey = sp.EndKey
+					} else {
+						completedSpan.Key = key
+						completedSpan.EndKey = nextKey
+					}
+
+					mu.Lock()
+					mu.completedSpans = append(mu.completedSpans, completedSpan)
+					mu.completedSpanIdx = append(mu.completedSpanIdx, idx)
+					mu.Unlock()
+
+					if knobs, ok := ibm.flowCtx.Cfg.TestingKnobs.IndexBackfillMergerTestingKnobs.(*IndexBackfillMergerTestingKnobs); ok {
+						if knobs != nil && knobs.PushesProgressEveryChunk {
+							pushProgress()
+						}
+					}
+
+					key = nextKey
+				}
+
+				if err != nil {
+					return err
+				}
+			}
+
+			return nil
+		})
+
+		err = g.Wait()
+	}()
+
+	for prog := range progCh {
+		p := prog
+		if p.CompletedSpans != nil {
+			log.VEventf(ctx, 2, "sending coordinator completed spans: %+v", p.CompletedSpans)
+		}
+
+		ibm.output.Push(nil, &execinfrapb.ProducerMetadata{BulkProcessorProgress: &p})
+	}
+
+	if err != nil {
+		ibm.output.Push(nil, &execinfrapb.ProducerMetadata{Err: err})
+	}
+}
+
+var _ execinfra.Processor = &IndexBackfillMerger{}
+
+// Merge merges the entries from startKey to endKey from the index with sourceID
+// into the index with destinationID, up to a maximum of chunkSize entries.
+func (ibm *IndexBackfillMerger) Merge(
+	ctx context.Context,
+	codec keys.SQLCodec,
+	table catalog.TableDescriptor,
+	sourceID descpb.IndexID,
+	destinationID descpb.IndexID,
+	startKey roachpb.Key,
+	endKey roachpb.Key,
+	chunkSize int64,
+) (roachpb.Key, error) {
+	sourcePrefix := rowenc.MakeIndexKeyPrefix(codec, table.GetID(), sourceID)
+	prefixLen := len(sourcePrefix)
+	destPrefix := rowenc.MakeIndexKeyPrefix(codec, table.GetID(), destinationID)
+
+	key := startKey
+	destKey := make([]byte, len(destPrefix))
+
+	if knobs, ok := ibm.flowCtx.Cfg.TestingKnobs.IndexBackfillMergerTestingKnobs.(*IndexBackfillMergerTestingKnobs); ok {
+		if knobs != nil && knobs.RunBeforeMergeChunk != nil {
+			if err := knobs.RunBeforeMergeChunk(key); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	err := ibm.flowCtx.Cfg.DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		// For now just grab all of the destination KVs and merge the corresponding entries.
+		kvs, err := txn.Scan(ctx, key, endKey, chunkSize)
```

Comment:
If I were being nit-picky, I'd encourage you to use the lower-level APIs. These high-level APIs are sort of garbage. I'd consider something like:

```
var ba roachpb.BatchRequest
	ba.TargetBytes = 16 << 10
	ba.MaxSpanRequestKeys = chunkSize
	ba.Add(&roachpb.ScanRequest{
		RequestHeader: roachpb.RequestHeader{
			Key:    key,
			EndKey: endKey,
		},
		ScanFormat: roachpb.KEY_VALUES,
	})
	br, pErr := txn.Send(ctx, ba)
	if pErr != nil {
		return pErr.GoError()
	}
	resp := br.Responses[0].GetScan()
	for _, row := range resp.Rows {
		
	}
```
Otherwise you're liable to pull a lot more data than you wanted.

Reply:
Happy to update to this now.  Was 16Kb just a bet guess or do you have some thinking there I can include in a comment?

Reply:
Decided to push this to a follow up PR. Will open it shortly after this merges.